### ISSUE TITLE
Statistics counters use atomics without padding (false sharing)

### ISSUE THREAD

**Issue by hhoffstaette** [PowerDNS/pdns#9451]
- Program: dnsdist but probably also Authoritative & Recursor
 - Issue type: Performance improvement (not really a functional defect)

### Not so short description
Poking around in dnsdist I saw that statistics, request/reply counters etc. are declared as std::atomic
(which is fine) but almost always declared without padding, aligned next to each other.
This leads to a symptom known as _false sharing_ and can have surprising performance impact
when several threads try to update atomics that are aligned next to each other. As they share
the same cacheline(s) they are constantly wrongly invalidated when one of the 'neighbours'
is updated. Since dnsdist (and I assume other PowerDNS producst as well) use quite a few threads
and aim for low latency, avoiding thread pile-up and burning instructions (by queueing the CAS operations)
is essential.

Occurrences of this are easy to spot e.g. in `DNSDistStats`, `ClientState` and `DownstreamState`.
`DNSDistStats` already conveniently declares a `stat_t = std::atomic<uint_64_t>` so adding
alignment to this type (and using the type in more places) should be a quick fix.

Normally I'd do this myself as PR but there are several ways to accomplish this alignment,
and I'd rather leave the decision to the maintainers. IMHO the easiest & most compatible way
would be to follow Google's Abseil / Facebook's Folly which you can find [here](https://github.com/abseil/abseil-cpp/blob/master/absl/base/optimization.h#L56).
Alternatively there's also std::alignas() and friends.

**Comment by rgacogne** [comment]
I have been meaning to benchmark the use of mcounter [1] for a while. For the record, @ahupowerdns mentioned on IRC that he did benchmark the performance impact of false sharing and did not find it very noticeable in our case.

[1]: https://github.com/ahupowerdns/mcounter

**Comment by hhoffstaette** [comment]
> @ahupowerdns mentioned on IRC that he did benchmark the performance impact of false sharing and did
> not find it very noticeable in our case.

Well, that's certainly true for *my* use case (home lab with 2 clients), but I know that people are using dnsdist as first line of defense with very high packet loads. Also note that atomic updates become progressively more expensive with the core count and doubly so on NUMA. It's a simple optimization that doesn't cost much and caught my eye while looking for something else (how to clear the statistics counters, issue #3459). mcounter looks interesting, thanks for poiting that out!

**Comment by rgacogne** [comment]
I totally agree with you that it would be easy to specify like `std::alignas(64)` (or more exactly something a bit smarter, like is done in the example you provided) on `DNSDistStats::stat_t` for a start, and generalizing the use of that type. That should be only beneficial, so I'll be happy to merge something like that, and we can investigate mcounter later :-)

**Comment by rgacogne** [comment]
(too bad we don't have `std::hardware_destructive_interference_size` in C++11).

**Comment by hhoffstaette** [comment]
So I looked at this and it turns out (of course) that alignment cannot be specified for declared types since `using` or `typedef` don't have storage. Understandably this upsets the compiler. :(
Clever ideas? IMHO simply adding the alignment wherever it's needed should be easiest and least error-prone.
Alternatively I can `#define` a counter_t but that will break the boost type accessors, so .. argh.

**Comment by hhoffstaette** [comment]
Let me know if the initial commit goes into an acceptable direction. It still compiles and works. :)

**Comment by hhoffstaette** [comment]
The platform's cacheline size can be queried via `getconf LEVEL1_DCACHE_LINESIZE`, so someone who knows autoconf could make it configurable. 64 seems to be a common default across many server platforms, even newer ARMs.

**Comment by zeha** [comment]
Note that not all platforms return a non-zero value for `getconf LEVEL1_DCACHE_LINESIZE`:

```
$ getconf LEVEL1_DCACHE_LINESIZE
0
$ uname -a
Linux abel 4.19.0-10-armmp-lpae #1 SMP Debian 4.19.132-1 (2020-07-24) armv7l GNU/Linux
```

**Comment by hhoffstaette** [comment]
> Note that not all platforms return a non-zero value for `getconf LEVEL1_DCACHE_LINESIZE`

No, of course not.. :cry: 
With the new defines you can at least inject the value if you know it.

**Comment by omoerbeek** [comment]
Are you proposing the build platform influences the alignment? I do not think that is a sane idea, the build machine might not have the same value as the runtime machine. Runtime there are possibilities, of course. e.g. via an aligned dynamic allocation.

**Comment by hhoffstaette** [comment]
> Are you proposing the build platform influences the alignment?
I'm saying that being able to specify the value is better than to hardcode a common value of 64; nothing more, nothing less. You can always build with 0 and get the old (unaligned) behaviour.
The point here was primarily to move towards a common type for stats instead of raw atomics so that it can be easily
changed later (see the note about mcounter ). If you want to completely rewrite everything to be runtime-detected, be my guest but I'm not doing that.

**Comment by omoerbeek** [comment]
I was triggered by your suggestion to use getconf configure time. I think such constructs should be avoided.

**Comment by omoerbeek** [comment]
https://gist.github.com/omoerbeek/a81d8f863bc8606a976a2028cd8ac470

shows how we could approach things. Drawback is still that the alignment must be a `constexpr`. If we want to avoid that (and be able to runtime determine the alignment) we could allocate using a dynamic allocation (e.g. posix_memalign(3) or a C++ equivalent if it exists)

**Comment by omoerbeek** [comment]
https://gist.github.com/omoerbeek/adab5e443582b102d6acb51fb7b2ceff shows an alternative approach that allows for runtime determination of the alignment..

**Comment by rgacogne** [comment]
Do we expect a case where the runtime value is different from the build time one? And if that happens, at worst we would end up in pretty much the same position as the one we are today if the value is too small, right?

**Comment by omoerbeek** [comment]
I assume the runtime value to be based on the CPU model. One CPU might have 32, another 128. e.g. If you happen to build on a 32 one and run on a 128, the 128 is going to suffer.

**Comment by zeha** [comment]
I think if this were a built-time setting, at least we'd want a cross-build compatible way of figuring out the value, because some downstreams (=Debian) seem to care about this.

**Comment by omoerbeek** [comment]
Did s bit of research, while theoretically possible, it seems that varying L1 cache line sizes between CPU models is extremely rare.

MariaDB uses this snippet:

```
/*
  Provide defaults for the CPU cache line size, if it has not been detected by
  CMake using getconf
*/

#if !defined(CPU_LEVEL1_DCACHE_LINESIZE) || CPU_LEVEL1_DCACHE_LINESIZE == 0
  #if defined(CPU_LEVEL1_DCACHE_LINESIZE) && CPU_LEVEL1_DCACHE_LINESIZE == 0
    #undef CPU_LEVEL1_DCACHE_LINESIZE
  #endif

  #if defined(__s390__)
    #define CPU_LEVEL1_DCACHE_LINESIZE 256
  #elif defined(__powerpc__) || defined(__aarch64__)
    #define CPU_LEVEL1_DCACHE_LINESIZE 128
  #else
    #define CPU_LEVEL1_DCACHE_LINESIZE 64
  #endif
#endif
```

**Comment by rgacogne** [comment]
Good, because I'd really much prefer something that doesn't involve aligning dynamic memory allocations if possible :-)

**Comment by omoerbeek** [comment]
https://github.com/omoerbeek/pdns/commit/72db5feea4750909b2246023489fe039901d5767

Only compilel tested!

**Comment by hhoffstaette** [comment]
> [omoerbeek@72db5fe](https://github.com/omoerbeek/pdns/commit/72db5feea4750909b2246023489fe039901d5767)
> 
> Only compile tested!

That's certainly much nicer than having to sprinkle the alignment all over the place. :+1:

**Comment by rgacogne** [comment]
Fixed by #9998.
